Basic guide for "best practices" and preparing to ship (#169)
replicatedhq/kots.io: content/vendor/guides/production-ready.md

---
date: "2020-02-20T00:00:00Z"
lastmod: "2020-02-20T00:00:00Z"
title: "Preparing for Production Deploys"
weight: "1009"
draft: false
---

This guide is intended as a lightweight checklist to ensure you have explored all the features of KOTS and integrated the ones that will help ensure your end users are most likely to be successful deploying and running your KOTS application. This guide is intended to grow as more features are integrated and documented in KOTS.

It is split into three parts:

- [Implementation Steps](#implementation-steps)
- [Further Reading](#further-reading)
- [Appendix A: A Questionnaire for your End Users](#appendix-a-a-questionnaire-for-your-end-users)

* * *

## Implementation Steps

### Basic Whitelabeling

> All applications should be whitelabeled with [a name and an icon](/reference/v1beta1/application/#title). 

Reviewer: ```suggestion
All applications should be whitelabeled with [a title and an icon](/reference/v1beta1/application/#title). 
```

If you expect to also install stateful services into existing clusters, you'll likely want to expose [preflight analyzers that check for the existence of a storage class](https://troubleshoot.sh/reference/analyzers/storage-class/).

### Namespaces

It is *strongly* advised that applications be architected to deploy a single application into a single namespace when possible. This will give the most flexibility when deploying to end user environments. Don't specify a namespace in your YAML resources, or try to make this user-configurable using the `kots.io` `Config` object, just leave namespace blank. Letting the end user manage namespaces will be the easiest way to reduce friction. The ability to manage namespaces in kotsadm will be added in an upcoming KOTS release, so if you intend to  deploy multiple apps using a single Admin Console, it is be reasonable to architect your app assuming the end user will manage the Namespaces that each component runs in.

### Helm

> Helm charts are supported by KOTS but not required. If for applications that are already packaged using helm, then the helm support in KOTS can help get an app packaged faster. If an application does not presently use helm, there's no requirement to use helm, as the KOTS built-in templating includes much of the same functionality, and the Admin Console includes a deep [kustomize.io integration](/vendor/kotsadm/updating/patching-with-kustomize) to greatly reduce the amount of templating required by app maintainers in the first place.

Reviewer: 🙌

### Operators
Operators are good for specific use cases, we've written in-depth about them in our [Operators Blog Post](https://blog.replicated.com/operators-in-kots/). In general we recommend thinking deeply about the problem space an application solves before going down the operator path. They're really cool and powerful, but take a lot of time to build and maintain

<!-- coming soon, wait for guide
### Licensing and Entitlements
  - how we think about it 
  - basic enforcement
  - advanced enforcement via API

Basic license enforcement comes baked in for free -- expiration 

Our recommendation is always to enforce licensing with a strong contract, but if you have strict licensing needs, you can always implement [Runtime License Validation](/vendor/guides/runtime-license-validation) in your application. This involves modifying your application code to query an API exposed by KOTS to consume the license payload, and validating the signature in the hot code paths in your application's core business logic.
-->

### Bundling and Analyzing Logs with Support bundle

> A robust support bundle is essential to minimizing back-and-forth when things go wrong. At a very minimum every app's support bundle should contain logs for an application's core pods. Usually this will be done with label selectors. To the labels for an application, either inspect the YAML, or run

Reviewer: ```suggestion
A robust support bundle is essential to minimizing back-and-forth when things go wrong. At a very minimum every app's support bundle should contain logs for an application's core pods. Usually this will be done with label selectors. To get the labels for an application, either inspect the YAML, or run
```

```shell
kubectl get pods --show-labels
```

> against a running instance to see what labels are used. Once the labels are discoved, a [logs collector](https://troubleshoot.sh/reference/collectors/pod-logs/) can be used to include logs from these pods in a bundle. Depending on the complexity of an app's labeling schema, you may need a few different declarations of the `logs` collector.

Reviewer: ```suggestion
against a running instance to see what labels are used. Once the labels are discovered, a [logs collector](https://troubleshoot.sh/reference/collectors/pod-logs/) can be used to include logs from these pods in a bundle. Depending on the complexity of an app's labeling schema, you may need a few different declarations of the `logs` collector.
```

> As commmon issues are encountered in the field, it will make sense to add not only collectors but also analyzers to an apps's support stack. For example, when an error in a log file is discoved that should be surfaced to an end user user in the future, a simple [Text Analyzer](https://troubleshoot.sh/reference/analyzers/regex/) can detect specific log line in the future and inform an end user of remediation steps.

Reviewer: ```suggestion
As common issues are encountered in the field, it will make sense to add not only collectors but also analyzers to an apps's support stack. For example, when an error in a log file is discovered that should be surfaced to an end user in the future, a simple [Text Analyzer](https://troubleshoot.sh/reference/analyzers/regex/) can detect specific log line in the future and inform an end user of remediation steps.
```

<!-- coming soon
For a full breakdown check out our [Support Bundle Guide](/vendor/guides/support-bundle)
-->

### Adding Prometheus Graphs

If an application exposes Prometheus metrics, we recommend integrating [Custom Graphs](/vendor/config/dashboard-graphs) to expose deep application metrics to end users.

### Building a Collaborative Workflow

We recommend using a git-based workflow, as presented in the [KOTS starter Repo](https://github.com/replicatedhq/replicated-starter-kots). This will allow teams to map git branches to channels in the Vendor Portal, and allow multiple team members to seamlessly collaborate across features and releases.

###### Tagging Releases for Production

> In addition to the starter Makefile included in the [KOTS starter Repo](https://github.com/replicatedhq/replicated-starter-kots), Replicated provides an [Advanced Makefile](https://github.com/replicatedhq/replicated-starter-kots/tree/master/Makefile.advanced) that builds on the basic one. It adds logic for making production releases using git tags. The basic outline is:

Reviewer: ```suggestion
In addition to the starter `Makefile` included in the [KOTS starter Repo](https://github.com/replicatedhq/replicated-starter-kots), Replicated provides an [Advanced Makefile](https://github.com/replicatedhq/replicated-starter-kots/tree/master/Makefile.advanced) that builds on the basic one. It adds logic for making production releases using git tags. The basic outline is:
```

- On pushes to the `master` branch, create a release on unstable with the name `Unstable-${SHA}`
- On pushing a git tag, create a release on the beta branch, using the name `Beta-${TAG}` for the release version.

> - Our recommendation is that these tags be tested, and then the release be manually promoted to the `Stable` channel using https://vendor.replicated.com. The goal in using manual promotion to restrict who can cause new versions to go out to users via RBAC roles in the Vendor Portal.

Reviewer: ```suggestion
- Our recommendation is that these tags be tested, and then the release be manually promoted to the `Stable` channel using [vendor.replicated.com](https://vendor.replicated.com). The goal in using manual promotion to restrict who can cause new versions to go out to users via RBAC roles in the Vendor Portal.
```
